Add an article for properties.

BillWagner opened the pull request:
Fixes #55 
Describe the basics of properties, and how to use them.

After PR #175 is merged, I'll add links to upcoming topics and rebase this article on those changes.

BillWagner committed:
Add an article for properties.

Describe the basics of properties, and how to use them.

Move properties article to new location
Remove wrench icon from completed articles.

cartermp commented on docs/languages/csharp-properties.md:
```diff
@@ -0,0 +1,253 @@
+        {
+            if (string.IsNullOrWhiteSpace(value))
+                throw new ArgumentException("First name must not be blank");
+                if (value != firstName)
```
Formatting here is off, making this `if()` statement appear as if it were a child of the above `if()` statement.  I'd also considering using braces just because the one-line conditional statement is long enough to warrant a line break.

stevehoag commented on docs/languages/csharp-properties.md:
```diff
@@ -0,0 +1,253 @@
+that accurately expresses their design intent.
+
+Properties behave like fields when they are accessed.
+However, unlinke fields, properties are implemented
```
"unlike"?

stevehoag commented on docs/languages/csharp-properties.md:
```diff
@@ -0,0 +1,253 @@
+public class Person
+{
+	public string FirstName;
+	// elided
```
I had to look up "elided" - not a common term. Repeated in numerous code examples. Do we even need a comment here, and has the reader already been introduced to the "//" comment syntax?

mairaw commented on docs/languages/csharp-properties.md:
```diff
@@ -0,0 +1,253 @@
+in your scenario: You can check the relationships between different properties, or validate
+against any external conditions. Any valid C# statements are valid in a property accessor.
+
+## INotifyPropertyChanged
```
This seems a more advanced scenario for someone who's just learning C# and what a property is. Should this come at least later in the topic?

mairaw commented on docs/languages/csharp-properties.md:
```diff
@@ -0,0 +1,253 @@
+## INotifyPropertyChanged
+
+Another common scenario where you need to write code in a property accessor is to
+support the `INotifyPropertyChanged` interface used in XAML based data binding.
```
Instead of mentioning XAML here, should we say something more generic like "used to notify binding clients that a property value has changed". Do we also need to explain what data binding is before explaining how the interface works?

BillWagner commented:
I have updated the article based on all the comments,
moved the properties article to be inline with the new structure (which did unlink the comments :sob: )
Added link to the Properties article in the C# index.

I believe this should be ready.

cartermp commented on docs/languages/csharp/properties.md:
````diff
@@ -0,0 +1,255 @@
+The syntax for properties is a natural extension to
+fields. A field defines a storage location:
+
+```
````
This sample and the others should have `cs` appended to the "```" so that we can get C# syntax highlighting:

```
```cs
// the code
```

mairaw commented on docs/languages/csharp/properties.md:
```diff
@@ -0,0 +1,255 @@
+
+Up to this point, all the property definitions you have seen are read-write properties
+with public accessors. That's not the only valid accessibility for properties.
+You can create readonly properties, or give different accessibility to the set and get
```
According to the Microsoft Manual of Style, read-only should always be hyphenated (unless you're talking about the keyword, of course). And it should be read/write instead of read-write.

cartermp commented on docs/languages/csharp/properties.md:
````diff
@@ -0,0 +1,255 @@
+}
+```
+
+The example above uses the new *String Interpolation* syntax to create a 
````
I'd avoid the use of "new" here.  We'd like these samples to reflect the latest and greatest way to do things except for in cases where we're providing a sample for an older version of .NET where Roslyn features can't be used.

cartermp commented on docs/languages/csharp/properties.md:
```diff
@@ -0,0 +1,255 @@
+The example above uses the new *String Interpolation* syntax to create a 
+the formatted string for the full name.
+
+Another new C# feature called *Expression Bodied Members* provides a more
```
Same feedback as above.

cartermp commented on docs/languages/csharp/properties.md:
````diff
@@ -0,0 +1,255 @@
+}
+```
+
+Some of the syntax shown above may be new to you. The `?.` operator is called
````
Same feedback as above regarding the use of "new" when describing the syntax.

cartermp commented on docs/languages/csharp/properties.md:
```diff
@@ -0,0 +1,255 @@
+Properties are a form of smart fields in a class or object. From
+outside the object, they appear like fields in the object. However,
+Properties can be implemented using the full palette of C# functionality.
+You can provide validation, different accessibility, lazy evaluation,
```
Regarding lazy evaluation: It would be nice to have an example of this above.

cartermp commented on docs/languages/csharp/properties.md:
```diff
@@ -0,0 +1,255 @@
+# Introduction
```
Let's have you list your name as the author.  Want to make sure you get the credit for the awesome guide!  Should be a link to your Github profile.  e.g.

Author: [Bill Wagner](https://github.com/BillWagner)

cartermp commented:
Left some feedback.  The article is excellent as a whole.  Thanks a lot!

mairaw commented on docs/languages/csharp/properties.md:
```diff
@@ -0,0 +1,255 @@
+the right side of the operator. The end result is that if there are no subscribers
+to the `PropertyChanged` event, the code to raise the event doesn't execute. It would
+throw a `NullReferenceException` without this check in that case. See the page on
+`events` for more details. This example also uses the new
```
should we add the link here?

mairaw commented on docs/languages/csharp/properties.md:
```diff
@@ -0,0 +1,255 @@
+throw a `NullReferenceException` without this check in that case. See the page on
+`events` for more details. This example also uses the new
+`nameof` operator to convert from the property name symbol to its text representation.
+Using `nameof` can reduce errors where you have mis-typed the name of the property.
```
mistyped instead?

mairaw commented on docs/languages/csharp/properties.md:
```diff
@@ -0,0 +1,255 @@
+## Computed Properties
+
+A property does not need to simply return the value of a member field. You can create properties
+that return a computed value. Let's expand the `Person` object to return the fulll name, computed
```
typo: full and not fulll

mairaw commented on docs/languages/csharp/properties.md:
````diff
@@ -0,0 +1,255 @@
+```
+
+The example above uses the new *String Interpolation* syntax to create a 
+the formatted string for the full name.
````
"to create the formatted" not "a the"

mairaw commented:
I agree with @cartermp. Excellent topic! Also provided some minor corrections.

matthewreily commented on docs/languages/csharp/properties.md:
```diff
@@ -0,0 +1,255 @@
+must have a void return type, you report errors in the set accessor by throwing an exception.
+
+That is a simple case of validation. You can extend this same syntax to anything needed
+in your scenario: You can check the relationships between different properties, or validate
```
I think scenario: should be scenario.

matthewreily commented:
Great job Bill!  Glad you are blazing the path for me to follow.

BillWagner commented:
I responded to all the feedback. (Thanks everyone). I believe this is ready.

mairaw commented on docs/languages/csharp/properties.md:
```diff
@@ -0,0 +1,345 @@
+The property definition shown above is a read-write property. Notice
+the keyword `value` in the set accessor. The `set` accessor always has
+a single parameter named `value`. The `get` accessor must return a value
+that is convertable to the type of the property (`string` in this example).
```
Overall everything LGTM! But one last edit perhaps: should this be convertible instead?

BillWagner replied:
Just updated.

matthewreily commented on docs/languages/csharp/properties.md:
```diff
@@ -0,0 +1,345 @@
+
+Properties are a form of smart fields in a class or object. From
+outside the object, they appear like fields in the object. However,
+Properties can be implemented using the full palette of C# functionality.
```
Properties should be lowercase **properties**

matthewreily commented:
LGTM minus on small capitalization nitpick.

BillWagner commented:
Just fixed @matthewreily 's last note on Properties v. properties.

cartermp commented:
LGTM.  Can you rebase to squash the PR feedback commits?

BillWagner commented:
@cartermp Done.

cartermp commented:
Thanks!
